### PR TITLE
fix ICE in ConstProp

### LINKED ISSUE #96169: ICE: "Got a scalar pair where a scalar was expected" -Zmir-opt-level=4  --emit=mir
### Code

```Rust
// run-rustfix
#![allow(unused)]
fn a() -> usize { 0 }
//~^ ERROR return types are denoted using `->`

fn bar(_: u32) {}

fn baz() -> *const dyn Fn(u32) { unimplemented!() }

fn foo() {
    match () {
        _ if baz() == &bar as &dyn Fn(u32) => (),
        () => (),
    }
}

fn main() {
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (878c7833f 2022-04-16)
binary: rustc
commit-hash: 878c7833f6c1ff10e2fd89074e5bd4ef5ff15936
commit-date: 2022-04-16
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.0
```

### Error output

```

```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/compiler/rustc_const_eval/src/interpret/operand.rs:77:42: Got a scalar pair where a scalar was expected

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/compiler/rustc_errors/src/lib.rs:1349:9
stack backtrace:
   0:     0x7f0edf49d7fd - std::backtrace_rs::backtrace::libunwind::trace::h533af29c806df9d1
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f0edf49d7fd - std::backtrace_rs::backtrace::trace_unsynchronized::hdb4c2ff88b52ef66
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f0edf49d7fd - std::sys_common::backtrace::_print_fmt::h68495a06222c4efb
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f0edf49d7fd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h91043f683408357d
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f0edf4f8eec - core::fmt::write::hcc15453d8963f001
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/core/src/fmt/mod.rs:1194:17
   5:     0x7f0edf48ef51 - std::io::Write::write_fmt::h5217846c280e1260
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/io/mod.rs:1655:15
   6:     0x7f0edf4a0515 - std::sys_common::backtrace::_print::h88be4bfce1061450
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f0edf4a0515 - std::sys_common::backtrace::print::he017a6566e010988
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f0edf4a0515 - std::panicking::default_hook::{{closure}}::hb793c4d82607bce7
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/panicking.rs:295:22
   9:     0x7f0edf4a0189 - std::panicking::default_hook::hc3c558385f484fdc
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/panicking.rs:314:9
  10:     0x7f0edfcd1421 - rustc_driver[19d608e2c8daf1f4]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f0edf4a0ce6 - std::panicking::rust_panic_with_hook::h28c3b2fc9e3a60cd
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/panicking.rs:702:17
  12:     0x7f0ee0d28881 - std[cd658a55527ed977]::panicking::begin_panic::<rustc_errors[9417c7ccc58903d8]::ExplicitBug>::{closure#0}
  13:     0x7f0ee0d27d96 - std[cd658a55527ed977]::sys_common::backtrace::__rust_end_short_backtrace::<std[cd658a55527ed977]::panicking::begin_panic<rustc_errors[9417c7ccc58903d8]::ExplicitBug>::{closure#0}, !>
  14:     0x7f0ee0cbddef - std[cd658a55527ed977]::panicking::begin_panic::<rustc_errors[9417c7ccc58903d8]::ExplicitBug>
  15:     0x7f0ee0c6f166 - std[cd658a55527ed977]::panic::panic_any::<rustc_errors[9417c7ccc58903d8]::ExplicitBug>
  16:     0x7f0ee0c6afe5 - <rustc_errors[9417c7ccc58903d8]::HandlerInner>::bug::<&alloc[c0e4610846cb0c5f]::string::String>
  17:     0x7f0ee0c6ad50 - <rustc_errors[9417c7ccc58903d8]::Handler>::bug::<&alloc[c0e4610846cb0c5f]::string::String>
  18:     0x7f0ee0d314bd - rustc_middle[77cffc0d0a9caf4f]::ty::context::tls::with_opt::<rustc_middle[77cffc0d0a9caf4f]::util::bug::opt_span_bug_fmt<rustc_span[ae9ba53627e4b7a2]::span_encoding::Span>::{closure#0}, ()>
  19:     0x7f0ee0d33f46 - rustc_middle[77cffc0d0a9caf4f]::util::bug::opt_span_bug_fmt::<rustc_span[ae9ba53627e4b7a2]::span_encoding::Span>
  20:     0x7f0ee0d33eb3 - rustc_middle[77cffc0d0a9caf4f]::util::bug::bug_fmt
  21:     0x7f0edfebeecf - <rustc_const_eval[2c45cfb8cbd0596a]::interpret::operand::Immediate>::to_scalar
  22:     0x7f0ee10cd5e2 - <rustc_mir_transform[30431bd415c4c272]::const_prop::ConstPropagator as rustc_middle[77cffc0d0a9caf4f]::mir::visit::MutVisitor>::visit_body
  23:     0x7f0ee10c9264 - <rustc_mir_transform[30431bd415c4c272]::const_prop::ConstProp as rustc_middle[77cffc0d0a9caf4f]::mir::MirPass>::run_pass
  24:     0x7f0ee110903e - rustc_mir_transform[30431bd415c4c272]::pass_manager::run_passes
  25:     0x7f0ee10f43de - rustc_mir_transform[30431bd415c4c272]::optimized_mir
  26:     0x7f0ee16f4720 - rustc_query_system[f27402853dfb6133]::query::plumbing::try_execute_query::<rustc_query_impl[604497a8fc516067]::plumbing::QueryCtxt, rustc_query_system[f27402853dfb6133]::query::caches::DefaultCache<rustc_span[ae9ba53627e4b7a2]::def_id::DefId, &rustc_middle[77cffc0d0a9caf4f]::mir::Body>>
  27:     0x7f0ee17e6d0e - <rustc_query_impl[604497a8fc516067]::Queries as rustc_middle[77cffc0d0a9caf4f]::ty::query::QueryEngine>::optimized_mir
  28:     0x7f0ee1bd753d - <rustc_middle[77cffc0d0a9caf4f]::ty::context::TyCtxt>::instance_mir
  29:     0x7f0ee0c9998f - rustc_middle[77cffc0d0a9caf4f]::mir::pretty::write_mir_pretty
  30:     0x7f0edff0bd10 - rustc_mir_transform[30431bd415c4c272]::dump_mir::emit_mir
  31:     0x7f0ee1d29b86 - <rustc_interface[c2767953942b9161]::queries::Queries>::ongoing_codegen
  32:     0x7f0ee1cf02fb - <rustc_interface[c2767953942b9161]::interface::Compiler>::enter::<rustc_driver[19d608e2c8daf1f4]::run_compiler::{closure#1}::{closure#2}, core[9a854ec5d686c1fc]::result::Result<core[9a854ec5d686c1fc]::option::Option<rustc_interface[c2767953942b9161]::queries::Linker>, rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>>
  33:     0x7f0ee1d18ecf - rustc_span[ae9ba53627e4b7a2]::with_source_map::<core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>, rustc_interface[c2767953942b9161]::interface::create_compiler_and_run<core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>, rustc_driver[19d608e2c8daf1f4]::run_compiler::{closure#1}>::{closure#1}>
  34:     0x7f0ee1cf0ee7 - <scoped_tls[673d87c3fad674a0]::ScopedKey<rustc_span[ae9ba53627e4b7a2]::SessionGlobals>>::set::<rustc_interface[c2767953942b9161]::interface::run_compiler<core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>, rustc_driver[19d608e2c8daf1f4]::run_compiler::{closure#1}>::{closure#0}, core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>>
  35:     0x7f0ee1d062ff - std[cd658a55527ed977]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[c2767953942b9161]::util::run_in_thread_pool_with_globals<rustc_interface[c2767953942b9161]::interface::run_compiler<core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>, rustc_driver[19d608e2c8daf1f4]::run_compiler::{closure#1}>::{closure#0}, core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>>::{closure#0}, core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>>
  36:     0x7f0ee1d06439 - <<std[cd658a55527ed977]::thread::Builder>::spawn_unchecked_<rustc_interface[c2767953942b9161]::util::run_in_thread_pool_with_globals<rustc_interface[c2767953942b9161]::interface::run_compiler<core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>, rustc_driver[19d608e2c8daf1f4]::run_compiler::{closure#1}>::{closure#0}, core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>>::{closure#0}, core[9a854ec5d686c1fc]::result::Result<(), rustc_errors[9417c7ccc58903d8]::ErrorGuaranteed>>::{closure#1} as core[9a854ec5d686c1fc]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  37:     0x7f0edf4aab93 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha55af26f35189599
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/alloc/src/boxed.rs:1866:9
  38:     0x7f0edf4aab93 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6a841b0f36efd076
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/alloc/src/boxed.rs:1866:9
  39:     0x7f0edf4aab93 - std::sys::unix::thread::Thread::new::thread_start::h4d154f7a3783bf9b
                               at /rustc/878c7833f6c1ff10e2fd89074e5bd4ef5ff15936/library/std/src/sys/unix/thread.rs:108:17
  40:     0x7f0edf2835c2 - start_thread
  41:     0x7f0edf308584 - __clone
  42:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.62.0-nightly (878c7833f 2022-04-16) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=4

query stack during panic:
#0 [optimized_mir] optimizing MIR for `foo`
end of query stack
error: aborting due to previous error
```

</p>
</details>

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/96169

### CODE REVIEW

**Comment by RalfJung** [comment]
I'm going to perf this to make sure the `track_caller` are not biting us.
@bors try @rust-timer queue

**Comment by estebank** [comment]
Code changes look good to me, r=me gated on perf but maybe @oli-obk wants to take a quick look.

**Comment by RalfJung** [comment]
@rust-timer build 0063a509fa6760dc55a40e101c02c17c02d0d6e7

**Review by oli-obk — APPROVED** [review]

**Comment by RalfJung** [comment]
Looks like somehow the try build failed. I thought test failures wouldn't be an issue, seemingly rustdoc failures *are* an issue.

@bors try @rust-timer queue

**Comment by RalfJung** [comment]
*oops* looks like we raced

File: compiler/rustc_const_eval/src/interpret/machine.rs [hunk 1]
```diff
@@ -10,7 +10,7 @@ use rustc_middle::mir;
 use rustc_middle::ty::{self, Ty, TyCtxt};
 use rustc_span::def_id::DefId;
 use rustc_target::abi::Size;
-use rustc_target::spec::abi::Abi;
+use rustc_target::spec::abi::Abi as FnAbi;
```
Comment (bjorn3) [inline compiler/rustc_const_eval/src/interpret/machine.rs:13]: `FnAbi` is already an existing type in `rustc_target::abi::call`. This might be confusing.

Comment (bjorn3) [reply compiler/rustc_const_eval/src/interpret/machine.rs:13]: Maybe SpecAbi?

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/machine.rs:13]: I don't under the term "spec" here at all. For me a "spec" is a "specification", aka defining preconditions and postconditions for the purpose of formal verification...

**Comment by RalfJung** [comment]
Wow, `track_caller` is really not cheap. I'll only add it for debug builds then.

**Comment by RalfJung** [comment]
This should be fine to roll up now.
@bors rollup-

**Pull request merged**